pkg/controller/jobs/rayjob/rayjob_controller.go

```diff
@@ -0,0 +1,279 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/controller/jobframework"
+)
+
+var (
+	gvk           = rayjobapi.GroupVersion.WithKind("RayJob")
+	FrameworkName = "ray.io/rayjob"
+)
+
+func init() {
+	utilruntime.Must(jobframework.RegisterIntegration(FrameworkName, jobframework.IntegrationCallbacks{
+		SetupIndexes:  SetupIndexes,
+		NewReconciler: NewReconciler,
+		SetupWebhook:  SetupRayJobWebhook,
+		JobType:       &rayjobapi.RayJob{},
+		AddToScheme:   rayjobapi.AddToScheme,
+	}))
+}
+
+// RayJobReconciler reconciles a Job object
+type RayJobReconciler jobframework.JobReconciler
+
+func NewReconciler(
+	scheme *runtime.Scheme,
+	client client.Client,
+	record record.EventRecorder,
+	opts ...jobframework.Option) jobframework.JobReconcilerInterface {
+	return (*RayJobReconciler)(jobframework.NewReconciler(scheme,
+		client,
+		record,
+		opts...,
+	))
+}
+
+type RayJob rayjobapi.RayJob
+
+func (j *RayJob) Object() client.Object {
+	return (*rayjobapi.RayJob)(j)
+}
+
+func (j *RayJob) IsSuspended() bool {
+	return j.Spec.Suspend
+}
+
+func (j *RayJob) IsActive() bool {
+	return j.Status.JobDeploymentStatus != rayjobapi.JobDeploymentStatusSuspended
+}
+
+func (j *RayJob) Suspend() {
+	j.Spec.Suspend = true
+}
+
+func (j *RayJob) ResetStatus() bool {
+	if j.Status.StartTime == nil {
+		return false
+	}
+	j.Status.StartTime = nil
+	return true
+}
+
+func (j *RayJob) GetGVK() schema.GroupVersionKind {
+	return gvk
+}
+
+func (j *RayJob) PodSets() []kueue.PodSet {
+	// len = workerGroups + head
+	podSets := make([]kueue.PodSet, len(j.Spec.RayClusterSpec.WorkerGroupSpecs)+1)
+
+	// head
+	podSets[0] = kueue.PodSet{
+		Name:     "head",
+		Template: *j.Spec.RayClusterSpec.HeadGroupSpec.Template.DeepCopy(),
+		Count:    1,
+	}
+
+	// workers
+	for index := range j.Spec.RayClusterSpec.WorkerGroupSpecs {
+		wgs := &j.Spec.RayClusterSpec.WorkerGroupSpecs[index]
+		replicas := int32(1)
+		if wgs.Replicas != nil {
+			replicas = *wgs.Replicas
+		}
+		podSets[index+1] = kueue.PodSet{
+			Name:     strings.ToLower(wgs.GroupName),
+			Template: *wgs.Template.DeepCopy(),
+			Count:    replicas,
+		}
+	}
+	return podSets
+}
+
+func applySelectors(dst, src map[string]string) map[string]string {
+	if len(dst) == 0 {
+		return src
+	}
+	for k, v := range src {
+		dst[k] = v
+	}
+	return dst
+}
+
+func (j *RayJob) RunWithNodeAffinity(nodeSelectors []jobframework.PodSetNodeSelector) {
+	j.Spec.Suspend = false
+	if len(nodeSelectors) == 0 {
+		return
+	}
+
+	// head
+	headPodSpec := &j.Spec.RayClusterSpec.HeadGroupSpec.Template.Spec
+	headPodSpec.NodeSelector = applySelectors(headPodSpec.NodeSelector, nodeSelectors[0].NodeSelector)
+
+	// workers
+	for index := range j.Spec.RayClusterSpec.WorkerGroupSpecs {
+		workerPodSpec := &j.Spec.RayClusterSpec.WorkerGroupSpecs[index].Template.Spec
+		workerPodSpec.NodeSelector = applySelectors(workerPodSpec.NodeSelector, nodeSelectors[index+1].NodeSelector)
```

Comment:
Better to check whether nodeSelector[index+1] is valid here. It's more safe and we may support partial admission in the future.